Experimental/amp skimlinks

slocka opened the pull request.

Commit:
Basic skim-js implementation

Commit:
Trying to use amp-analytics to send the request.

Commit:
Force tracking request to be a GET request

Commit:
add plant uml graph for design review

Commit:
Updating design doc graph

Commit:
design doc v2

Commit:
Start implementing AMP solution

Commit:
Hook AffiliateLinks to navigation click handler

Commit:
Better version of affiliate links manager.

Commit:
Adding tracking module & refactor booting of the extension

Commit:
Adding options and handle no links to resolve on the page scenario. (We always want to call beacon)

Commit:
Fix issue with wrong 'this' context

Commit:
Use events instead of optional callback to trigger na click tracking

Commit:
New link rewriter service

Commit:
Remove typo

Commit:
session id is not a thing anymore

Commit:
FIx JS Doc lint issues && move link rewrite service to 'src'

Commit:
Add domain resolver tests around beacon API calls

Commit:
Add domain resolver tests && replace AnchorRewriteData by createAnchorReplacementTuple

Commit:
adding test around beacon callback

Commit:
Add placeholder tests around waypoint query params

Commit:
Add impression id to waypoint params

Commit:
Implement .skip() tests

Commit:
Moving domain resolver tests to separate file.

Commit:
Add tracking tests

Commit:
No need to mock internal functions.

Commit:
Adding some NA click tracking tests

Commit:
Adding xcust test for NA click tracking

Commit:
Add test for na tracking => rnd query params

Commit:
Adding generate page impression id test

Commit:
Add tracking "setup" tests

Commit:
Cleaning up

Commit:
Adding tests  amp-skimlinks bootstrap

Commit:
Test beacon callback handler

Commit:
Add test to check if options are parsed when loading the extension

Commit:
Refactor link-rewrite-service priority logic.

Commit:
Add more link-rewriter-service tests

Commit:
Remove unused function

Commit:
Add tests for LinkRewriter

Commit:
Replace AnchorRewriteDataResponse by twoStepsResponse object.

Commit:
Add new variable to whitelist

Commit:
Fix use of .calledOnce instead of .called when testing not called.

Commit:
Use fakeWin instead of realWin for quicker run of the tests

Commit:
Add tests for rewriteAnchorUrl method

Commit:
Adding tests for link rewriter

Commit:
Refactor call beacon fallback when no links on page.

Commit:
Fix issue, missing anchor param for NA click

Commit:
Fix issue with impression tracking getting wrong ae status.

We need to make sure the anchorLinkReplacment map is updated before calling this.trackingService.sendImpressionTracking.

Commit:
amp-skimlinks should be loaded with layout='nodisplay'.

Commit:
Amp analytics requests should be sent using sendBeacon

Commit:
Consider 'canonical url hostname and sourceUrl hostname as 'internal' domain. (Domain will be excluded by default)

Commit:
Add real links for manual testing

Commit:
Renaming variables

Commit:
Refactor again the beacon callback logic so we can access the guid before sending  impression tracking.

Commit:
Remove .only

Commit:
Cleanup

Commit:
Call onDomUpdate from registerLinkRewriter instead of LinkRewriter constructor.

Commit:
Send new 'platform' flag instead of xcreo .

Commit:
Move LinkRewriter to be global to the app.

Commit:
Handle link-rewrite-service click handler from navigation instead of propagating a click event.

Commit:
Move constant to constants.js

Commit:
Renaming constants

Commit:
Renaming file

caccialdo commented on extensions/amp-skimlinks/0.1/affiliate-link-resolver.js:
```diff
@@ -0,0 +1,264 @@
+export const STATUS__AFFILIATE = 'affiliate';
+export const STATUS__NON_AFFILIATE = 'non-affiliate';
+export const STATUS__IGNORE_LINK = 'ignore';
+export const STATUS__UNKNOWN = 'unknown';
```
A short comment to explain what those are for unfamiliar peeps

caccialdo commented on extensions/amp-skimlinks/0.1/affiliate-link-resolver.js:
```diff
@@ -0,0 +1,264 @@
+   * @return {Promise}
+   */
+  resolveUnknownAnchors(anchorList) {
+    const alreadyResolved = this.mapToAnchorReplacementTuple_(anchorList);
```
`alreadyResolvedTuple` would be more explicit IMO

Commit:
Remove first param since the guid is passed using  setTrackingInfo

caccialdo commented on extensions/amp-skimlinks/0.1/affiliate-link-resolver.js:
```diff
@@ -0,0 +1,264 @@
+  /**
+   *
+   * @param {*} anchorList
+   * @return {Promise}
```
Returns an object

caccialdo commented on extensions/amp-skimlinks/0.1/affiliate-link-resolver.js:
```diff
@@ -0,0 +1,264 @@
+      // Set domains to STATUS__UNKNOWN to mark them as already requested.
+      this.markDomainsAsUnknown(domainsToAsk);
+      // Get anchors waiting for the API response to be resolved.
+      const pendingAnchors = this.getPendingAnchors_(anchorList, domainsToAsk);
```
`anchorsToDomainsToAsk` ?

caccialdo commented on extensions/amp-skimlinks/0.1/affiliate-link-resolver.js:
```diff
@@ -0,0 +1,264 @@
+   * @param {*} anchorList
+   */
+  mapToAnchorReplacementTuple_(anchorList) {
+    const anchorListNormalised = anchorList.map(anchor => {
```
`anchorReplacementTuple` ?

caccialdo commented on extensions/amp-skimlinks/0.1/affiliate-link-resolver.js:
```diff
@@ -0,0 +1,264 @@
+    return domains.reduce(((acc, domain) => {
+      const isResolved = this.domains_[domain];
+      const isExcluded = this.isExcludedDomain_(domain);
+      const isDuplicate = acc.indexOf(domain) !== -1;
```
Don't we have `[].includes()` in the AMP browser support?

slocka replied:
Not for arrays no.

caccialdo commented on extensions/amp-skimlinks/0.1/affiliate-link-resolver.js:
```diff
@@ -0,0 +1,264 @@
+  markDomainsAsUnknown(domains) {
+    domains.forEach(domain => {
+      const domainStatus = this.domains_[domain];
+      if (domainStatus && domainStatus !== STATUS__UNKNOWN) {
```
Can be simplified to `if (domainStatus) { return }` since we don't need to do anything too if `domainStatus === STATUS__UNKNOWN`

caccialdo commented on extensions/amp-skimlinks/0.1/affiliate-link-resolver.js:
```diff
@@ -0,0 +1,264 @@
+   * @param {string[]} domainsToAsk
+   */
+  getPendingAnchors_(anchorList, domainsToAsk) {
+    return anchorList.reduce((acc, anchor) => {
```
This is a filter operation really

caccialdo commented on extensions/amp-skimlinks/0.1/affiliate-link-resolver.js:
```diff
@@ -0,0 +1,264 @@
+   * @param {*} allDomains
+   * @param {*} affiliateDomains
+   */
+  updateDomainsStatusMapPostFetch_(allDomains, affiliateDomains) {
```
The `PostFetch` suffix can be removed as discussed offline.

caccialdo commented on extensions/amp-skimlinks/0.1/affiliate-link-resolver.js:
```diff
@@ -0,0 +1,264 @@
+    const beaconUrl = `${DOMAIN_RESOLVER_API_URL}?data=${JSON.stringify(data)}`;
+    const fetchOptions = {
+      method: 'GET',
+      // Disabled AMP CORS for dev
```
Is this something that will need to be changed before the final PR?

caccialdo commented on extensions/amp-skimlinks/0.1/affiliate-link-resolver.js:
```diff
@@ -0,0 +1,264 @@
+   */
+  getLinkDomain(anchor) {
+    let {hostname} = parseUrlDeprecated(anchor.href);
+    if (hostname.indexOf('www.') === 0) {
```
Similarly to `.includes()`, should we use `.startsWith()` here?

caccialdo commented on extensions/amp-skimlinks/0.1/affiliate-link-resolver.js:
```diff
@@ -0,0 +1,264 @@
+      hostname = hostname.replace('www.', '');
+    }
+
+    return hostname;
```
In fact the whole thing can be replaced by

```js
return hostname.replace(/^www\./, '');
```

caccialdo commented on extensions/amp-skimlinks/0.1/amp-skimlinks.js:
```diff
@@ -0,0 +1,145 @@
+  buildCallback() {
+    this.hasCalledBeacon = false;
+    this.xhr_ = Services.xhrFor(this.win);
+    this.ampDoc_ = this.getAmpDoc();
```
Bit unrelated to your PR but do you know why the document needs to be retrieved via a getter whereas the window is directly accessible on `this`?

If this is intentional, it could be that `this.getAmpDoc()` should always be called whenever it is used instead of storing a reference at initialisation time.

slocka replied:
Looks like it's just a shortcut getter for this.element.ampdoc_.

caccialdo commented on extensions/amp-skimlinks/0.1/amp-skimlinks.js:
```diff
@@ -0,0 +1,145 @@
+   * Send the impression tracking once we have the beacon API response.
+   */
+  onPageScanned_() {
+    let onBeaconApiResponse = this.affiliateLinkResolver.firstRequest;
```
```js
let beaconApiResponse = this.affiliateLinkResolver.firstRequest || this.affiliateLinkResolver.fetchDomainResolverApi([])
```

feels easier to read IMO

caccialdo commented on extensions/amp-skimlinks/0.1/skim-options.js:
```diff
@@ -0,0 +1,116 @@
+
+  const excludedDomainsAttr = element.getAttribute('excluded-domains');
+  if (excludedDomainsAttr) {
+    excludedDomains = excludedDomainsAttr.split(',');
```
I would trim the elements here, just to be on the safe side in case devs write `excluded-domains="a, b, c"`

```js
.split(',').map(domain => domain.trim())
```

Commit:
Extract waypoint logic from AffiliateLinkResolver

caccialdo commented on extensions/amp-skimlinks/0.1/tracking.js:
```diff
@@ -0,0 +1,212 @@
+   * @param {*} newInfo
+   */
+  setTrackingInfo(newInfo) {
+    this.trackingInfo_ = Object.assign(this.trackingInfo_, newInfo);
```
The assignment isn't required here since `Object.assign()` mutates the first argument

caccialdo commented on extensions/amp-skimlinks/0.1/tracking.js:
```diff
@@ -0,0 +1,212 @@
+      pref: referrer,
+      site: 'false',
+      url: anchor.href,
+      custom: anchor.getAttribute(XCUST_ATTRIBUTE_NAME) || customTrackingId,
```
Should there be an extra `|| null` in case there's no `customTrackingId` ?

slocka replied:
No, having a value of undefined removes the attributes from the queryString which is what we want.

caccialdo commented on extensions/amp-skimlinks/0.1/tracking.js:
```diff
@@ -0,0 +1,212 @@
+      pref: referrer,
+      uc: customTrackingId,
+      t: 1,
+    }, commonData);
```
We should probably hand pick the props from `commonData` to avoid copying random stuff. Not a massive deal though

slocka replied:
It's not random stuff, it's the data common to both page & link impression request :)
We want to copy everything on purpose.

caccialdo commented on extensions/amp-skimlinks/0.1/tracking.js:
```diff
@@ -0,0 +1,212 @@
+          count: 1,
+          ae: replacementUrl ? 1 : 0,
+        };
+      }
```
Perhaps simpler:

```js
urls[anchor.href] = urls[anchor.href] || {
    ae: replacementUrl ? 1 : 0,
    count: 0
}
urls[anchor.href].count += 1
```

caccialdo commented on extensions/amp-skimlinks/0.1/utils.js:
```diff
@@ -0,0 +1,41 @@
+/**
+ * Promise version of setTimeout(cb, 0);
+ */
+export function nextTick() {
```
💥

caccialdo commented on src/service/link-rewrite/link-rewrite-helpers.js:
```diff
@@ -0,0 +1,66 @@
+}
+
+/**
+ * Contains an optional synchronous and an optional asynchronous response.
```
Probably needs a comment about the reason for its existence since it's not a very frequent thing to see.

caccialdo commented on src/service/link-rewrite/link-rewrite-helpers.js:
```diff
@@ -0,0 +1,66 @@
+ *
+ * @param {*} twoStepsResponse
+ */
+export function isTwoStepsResponse(twoStepsResponse) {
```
As discussed offline, this could be relaxed to only check that at least one of the two props is set.

caccialdo commented on src/service/link-rewrite/link-rewrite-service.js:
```diff
@@ -0,0 +1,128 @@
+    const meta = this.iframeDoc_.querySelector(metaTagSelector);
+
+    if (meta && meta.hasAttribute('content')) {
+      this.priorityList_ = meta.getAttribute('content').trim().split(/\s+/);
```
Should we be consistent with `excluded-domains` and go for comma-separated strings? Or the opposite: space-separated everywhere.

Also, inverting the trim and split operations like `.split(/\s+/).map(sel => sel.trim())` would make it slightly more fault tolerant.

slocka replied:
I don't think it would since (/\s+) matches multiple spaces in a row.

caccialdo commented on src/service/link-rewrite/link-rewrite-service.js:
```diff
@@ -0,0 +1,128 @@
+   * @param {*} anchor -
+   */
+  maybeRewriteLink(anchor) {
+    const suitableLinkRewriters = this.getSuitableLinkRewritersForLink_(anchor);
```
Might be a stupid question but why not get the output of `this.getSuitableLinkRewritersForLink_()` already ordered by priority ?

caccialdo commented on src/service/link-rewrite/link-rewrite-service.js:
```diff
@@ -0,0 +1,128 @@
+      }
+      const replacedBy = chosenLinkRewriter ? chosenLinkRewriter.name : null;
+      suitableLinkRewriters.forEach(linkRewriter => {
+        linkRewriter.events.send(EVENTS.CLICK, {replacedBy, anchor});
```
Can't think of a better name: `linkRewriterId` ?

caccialdo commented on src/service/link-rewrite/link-rewriter.js:
```diff
@@ -0,0 +1,176 @@
+   * @param {*} options
+   */
+  constructor(iframeDoc, name, resolveUnknownLinks, options) {
+    this.name = name;
```
`id` might be better and consistent with other parts of the class.

caccialdo commented on src/service/link-rewrite/link-rewriter.js:
```diff
@@ -0,0 +1,176 @@
+      return null;
+    }
+
+    return this.anchorReplacementMap_.get(anchor);
```
`return this.anchorReplacementMap_.get(anchor) || null` to be consistent with the early return above.

slocka replied:
That wouldn't work if an extension wants to replace a url by an empty string. I think we can return whatever the extension replace value is.

caccialdo commented on src/service/link-rewrite/link-rewriter.js:
```diff
@@ -0,0 +1,176 @@
+   */
+  getReplacementUrl(anchor) {
+    if (!this.isWatchingLink(anchor)) {
+      return null;
```
Could returning `anchor.href` cause problems?

slocka replied:
I think it would work but it's a bit more obvious to not return anything if you are not planing to do any replacement

caccialdo commented on src/service/link-rewrite/link-rewriter.js:
```diff
@@ -0,0 +1,176 @@
+      // Mark all new anchors discovered to the default unknown.
+      // Note: Only anchors with a status will be considered in the click handlers.
+      // (Other anchors are assumed to be the ones exluded by linkSelector_)
+      this.updateAnchorMap_(unknownAnchors.map(anchor => {
```
```js
const unknownAnchorsTuples = unknownAnchors.map(createAnchorReplacementTuple)
this.updateAnchorMap_(unknownAnchorsTuples)
```

slightly better readability

caccialdo commented on src/service/link-rewrite/link-rewriter.js:
```diff
@@ -0,0 +1,176 @@
+    const unknownAnchors = this.getNewAnchors_(anchorList);
+
+    //  Ask for the affiliate status of the new anchors.
+    if (unknownAnchors.length) {
```
🔖 Invert `if` condition to remove the extra indentation

caccialdo commented on src/service/link-rewrite/link-rewriter.js:
```diff
@@ -0,0 +1,176 @@
+   * that were not in the page at the time of the last page scan.
+   * @param {*} anchorList
+   */
+  getNewAnchors_(anchorList) {
```
`getUnknownAnchors_` ?

Commit:
We can not use replacement macro for waypoint URL.

Macro in links only support CLIENT_ID and QUERY_PARAM.
https://github.com/ampproject/amphtml/blob/master/spec/amp-var-substitutions.md#variable-substitution-in-links

Commit:
Code review

Commit:
Clean up

Commit:
Add tests for link rewriter called from navigation.js

Commit:
Add basic validator

Commit:
Fixing check types

Commit:
Adding documentation for affiliate-link-resolver class

Commit:
Adding documentation for Tracking module

Commit:
Use skimresources instead of go.redirectingat

Commit:
Add documentation for link-rewriter

Commit:
Adding 1st version of the amp-skimlinks readme

Commit:
Fix broken tests

Commit:
Do not send NA click tracking on right click.

Commit:
Renaming test

Commit:
Fix typos

Commit:
Merge pull request #3 from skimhub/move-getWaypointUrl

Move get waypoint url

Commit:
Remove anchorClickActions since we don't need it after all

Commit:
Fix search & replace

pedroabreu commented on extensions/amp-skimlinks/0.1/affiliate-link-resolver.js:
```diff
@@ -0,0 +1,227 @@
+   * Build list of AnchorReplacementTuple
+   * @param {*} anchorList
+   */
+  mapToAnchorReplacementTuple_(anchorList) {
```
Random Q: Does the hanging `_` mean that it's a private method ? If so, do they follow any convention on it's order in a class (Either all private first or last) ?

There doesn't seem to be any particular order here which I guess answers my question 🙈

slocka replied:
Yes _ means it's private, I have added more jsdoc about it in an other PR. I also moved functions around to put the public methods at the top but it's my own convention I don't think they have a rule.

pedroabreu commented on extensions/amp-skimlinks/0.1/tracking.js:
```diff
@@ -0,0 +1,216 @@
+   * @param {*} startTime
+   */
+  sendImpressionTracking(anchorStatusMap, startTime) {
+    if (!this.tracking_) {
```
I guess this happens when tracking is disabled, perhaps a more readable way to have this is having a `shouldNotTrack_` (or `willNotTrack_`) method that reads the `this.tracking_` (Or any other future condition).

Not too fussed about it, just think it makes it easier to read/understand

pedroabreu commented on src/services.js:
```diff
@@ -251,6 +253,14 @@ export class Services {
+   * @return {!./service/link-rewrite/link-rewrite-service.LinkRewriteService}
+   */
+  static linkRewriteServiceForDoc(elementOrAmpDoc) {
+    return /** @type {!./service/link-rewrite/link-rewrite-service.LinkRewriteService} */ (
```
Is this correct ? 🤔Can the jsDoc be moved to the line above ?

pedroabreu commented on extensions/amp-skimlinks/0.1/skim-options.js:
```diff
@@ -0,0 +1,116 @@
+function getLinkSelector_(element) {
+  const linkSelector = element.getAttribute('link-selector');
+
+  return linkSelector || null;
```
Is the `null` fallback to prevent empty string ?

pedroabreu commented:
👏 👏 👏

Commit:
Fix issue when unregistered LinkRewriter are in priority list.

Commit:
Documentation fixes

Commit:
Rename link-rewrite-service to link-rewriter-manager for clarity

Commit:
Rename link-rewrite-service to link-rewriter-manager for clarity

Commit:
Fix new tests after rebase

Commit:
Replace Map by LinkReplacementCache class  since Map are not allowed in this project.

Commit:
Code review

Commit:
.trim() is already done in code above.

Commit:
Update pubcode for testing

Commit:
Merge pull request #4 from skimhub/add-jsdoc

Add jsdoc

Commit:
Code review

Commit:
Merge pull request #5 from skimhub/fixes

Fixes

Commit:
Fix broken tests

Commit:
Simplify updateLinkList

Commit:
FIx typo

Commit:
Add comment

Commit:
Merge pull request #6 from skimhub/link-replacement-cache

Link replacement cache

Commit:
Merge branch 'integration/amp-skimlinks-0.1' into experimental/amp-skimlinks
